lib/ansible/plugins/inventory/aws_ec2.py

```diff
@@ -0,0 +1,635 @@
+
+    def _format_instance_data(self, region, instances):
+        '''
+            :param region: the region in which the instance exists
+            :param instances: the reservations returned by boto3's describe_instances
+
+            A generator that yields named tuples with the fields region, instance_meta, and instance_data.
+        '''
+        instance_meta = {'Groups': instances.get('Groups', []),
+                         'OwnerId': instances.get('OwnerId'),
+                         'RequesterId': instances.get('RequesterId'),
+                         'ReservationId': instances.get('ReservationId')}
+        for instance_data in instances.get('Instances', []):
+            instance = namedtuple('instance', ['region', 'instance_meta', 'instance_data'])
+            instance.region = region
+            instance.instance_meta = instance_meta
+            instance.instance_data = instance_data
+            yield instance
+
+    def _get_instances_by_region(self, regions, filters, strict_permissions):
+        '''
+           :param regions: a list of regions in which to describe instances
+           :param filters: a list of boto3 filter dicionaries
+           :param strict_permissions: a boolean determining whether to fail or ignore 403 error codes
+           :return A list of namedtuples containing the fields region, instance_meta, and instance_data
+        '''
+        all_instances = []
+
+        for connection, region in self._boto3_conn(regions):
+            try:
+                paginator = connection.get_paginator('describe_instances')
+                instances = paginator.paginate(Filters=filters).build_full_result().get('Reservations')
+                if instances:
+                    instances = instances[0]
+            except botocore.exceptions.ClientError as e:
+                if e.response['ResponseMetadata']['HTTPStatusCode'] == 403 and not strict_permissions:
+                    instances = []
+                else:
+                    raise AnsibleError("Failed to describe instances: %s" % to_native(e))
+            except botocore.exceptions.BotoCoreError as e:
+                raise AnsibleError("Failed to describe instances: %s" % to_native(e))
+
+            if instances:
+                all_instances.extend(self._format_instance_data(region, instances))
+
+        return sorted(all_instances, key=lambda x: x.instance_data['InstanceId'])
+
+    def _get_hostname(self, instance, hostnames):
+        '''
+            :param instance: a named tuple with instance_data field
+            :param hostnames: a list of hostname destination variables in order of preference
+            :return the preferred identifer for the host
+        '''
+        if not hostnames:
+            hostnames = ['dns-name', 'private-dns-name']
+
+        for preference in hostnames:
+            hostname = self._get_boto_attr_chain(preference, instance)
+            if hostname:
+                if ':' in to_text(hostname):
```

Comment:
Would it be better to just always run hostnames through the munger? Seems like we could get back other bad values besides just `:` (though I don't remember if it will munge things in other ways like `-` to `_` that might cause problems).